subsys/net/lib/mqtt_sock/mqtt.c

```diff
@@ -0,0 +1,639 @@
+
+	tx_buf_init(client, &packet);
+
+	err_code = verify_tx_state(client);
+	if (err_code < 0) {
+		goto error;
+	}
+
+	err_code = subscribe_encode(param, &packet);
+	if (err_code < 0) {
+		goto error;
+	}
+
+	err_code = client_write(client, packet.cur, packet.end - packet.cur);
+
+error:
+	MQTT_TRC("[CID %p]:[State 0x%02x]: << result 0x%08x",
+		 client, client->internal.state, err_code);
+
+	mqtt_mutex_unlock(client);
+
+	return err_code;
+}
+
+int mqtt_unsubscribe(struct mqtt_client *client,
+		     const struct mqtt_subscription_list *param)
+{
+	int err_code;
+	struct buf_ctx packet;
+
+	NULL_PARAM_CHECK(client);
+	NULL_PARAM_CHECK(param);
+
+	mqtt_mutex_lock(client);
+
+	tx_buf_init(client, &packet);
+
+	err_code = verify_tx_state(client);
+	if (err_code < 0) {
+		goto error;
+	}
+
+	err_code = unsubscribe_encode(param, &packet);
+	if (err_code < 0) {
+		goto error;
+	}
+
+	err_code = client_write(client, packet.cur, packet.end - packet.cur);
+
+error:
+	mqtt_mutex_unlock(client);
+
+	return err_code;
+}
+
+int mqtt_ping(struct mqtt_client *client)
+{
+	int err_code;
+	struct buf_ctx packet;
+
+	NULL_PARAM_CHECK(client);
+
+	mqtt_mutex_lock(client);
+
+	tx_buf_init(client, &packet);
+
+	err_code = verify_tx_state(client);
+	if (err_code < 0) {
+		goto error;
+	}
+
+	err_code = ping_request_encode(&packet);
+	if (err_code < 0) {
+		goto error;
+	}
+
+	err_code = client_write(client, packet.cur, packet.end - packet.cur);
+
+error:
+	mqtt_mutex_unlock(client);
+
+	return err_code;
+}
+
+int mqtt_abort(struct mqtt_client *client)
+{
+	mqtt_mutex_lock(client);
+
+	NULL_PARAM_CHECK(client);
+
+	if (client->internal.state != MQTT_STATE_IDLE) {
+		client_disconnect(client, -ECONNABORTED);
+	}
+
+	mqtt_mutex_unlock(client);
+
+	return 0;
+}
+
+int mqtt_live(struct mqtt_client *client)
+{
+	u32_t elapsed_time;
+
+	NULL_PARAM_CHECK(client);
+
+	mqtt_mutex_lock(client);
+
+	if (MQTT_HAS_STATE(client, MQTT_STATE_DISCONNECTING)) {
+		client_disconnect(client, 0);
+	} else {
+		elapsed_time = mqtt_elapsed_time_in_ms_get(
+					client->internal.last_activity);
+
+		if ((MQTT_KEEPALIVE > 0) &&
+		    (elapsed_time >= (MQTT_KEEPALIVE * 1000))) {
+			(void)mqtt_ping(client);
+		}
+	}
+
+	mqtt_mutex_unlock(client);
+
+	return 0;
+}
+
+int mqtt_input(struct mqtt_client *client)
+{
+	int err_code = 0;
+
+	NULL_PARAM_CHECK(client);
+
+	mqtt_mutex_lock(client);
+
+	MQTT_TRC("state:0x%08x", client->internal.state);
+
+	if (MQTT_HAS_STATE(client, MQTT_STATE_DISCONNECTING)) {
+		client_disconnect(client, 0);
+	} else if (MQTT_HAS_STATE(client, MQTT_STATE_TCP_CONNECTED)) {
+		err_code = client_read(client);
+	} else {
+		err_code = -EACCES;
+	}
+
+	mqtt_mutex_unlock(client);
+
+	return err_code;
+}
+
+int mqtt_read_publish_payload(struct mqtt_client *client, void *buffer,
+			      size_t length)
+{
+	int ret;
+
+	NULL_PARAM_CHECK(client);
+
+	mqtt_mutex_lock(client);
+
+	if (client->internal.remaining_payload == 0) {
+		ret = 0;
+		goto error;
```

Comment:
Please-please let's change this to "goto out", or "done", or "finish", just not "error".